k8s/pkg/registryserver/registry_cache.go

```diff
@@ -0,0 +1,126 @@
+package registryserver
+
+import (
+	"fmt"
+	"github.com/networkservicemesh/networkservicemesh/k8s/pkg/apis/networkservice/v1"
+	nsmClientset "github.com/networkservicemesh/networkservicemesh/k8s/pkg/networkservice/clientset/versioned"
+	"github.com/networkservicemesh/networkservicemesh/k8s/pkg/networkservice/informers/externalversions"
+	"github.com/networkservicemesh/networkservicemesh/k8s/pkg/registryserver/resource_cache"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"time"
+)
+
+type RegistryCache interface {
+	AddNetworkService(ns *v1.NetworkService) (*v1.NetworkService, error)
+	GetNetworkService(name string) (*v1.NetworkService, error)
+
+	AddNetworkServiceManager(nsm *v1.NetworkServiceManager) (*v1.NetworkServiceManager, error)
+	GetNetworkServiceManager(name string) (*v1.NetworkServiceManager, error)
+
+	AddNetworkServiceEndpoint(nse *v1.NetworkServiceEndpoint) (*v1.NetworkServiceEndpoint, error)
+	DeleteNetworkServiceEndpoint(endpointName string) error
+	GetNetworkServiceEndpoints(networkServiceName string) []*v1.NetworkServiceEndpoint
+
```

Comment:
Hmm can this interface be implemented in a resource neutral way. Something like:

```go
type RegistryCache interface {
		AddEntry(nse interface{}) (interface{}, error)
		DeleteEntryByName(name string) error
		GetEntryByName(name string) []*interface{}
                Start()
                Stop()
}
```

Then each implementation will cast the `interface{}` to whatever it needs to.

Reply:
There is class `abstractResourceCache`. It implements all required logic in resource neutral way. Other classes - `NetworkServiceCache`, `NetworkServiceEndpointCache` and `NetworkServiceManagerCache` are concrete resource caches. They use `abstractResourceCache` inside and do required casts. `RegistryCache` - simple aggregation of all custom resource caches. It starts resource caches in transactional way and delegate Add/Get/Delete methods to appropriate resource cache.